File: apps/ship/src/execer/server-pool.ts
import { SshServer, type SshTarget } from "./ssh-server.js";

/** Reuses one SshServer per target key for the lifetime of a deploy step. */
export class ServerPool {
  private readonly servers = new Map<string, SshServer>();

  get(target: SshTarget): SshServer {
    const existing = this.servers.get(target.key);
    if (existing !== undefined) {
      return existing;
    }
    const created = new SshServer(target);
    this.servers.set(target.key, created);
    return created;
  }

  close(): void {
    for (const s of this.servers.values()) {
      s.close();
    }
    this.servers.clear();
  }
}

File: apps/nest/execer/server_pool.go
package execer

func NewServerPool() *ServerPool {
	return &ServerPool{}
}

type ServerPool struct {
	servers map[string]*Server
}

func (p *ServerPool) New(server *Server) *Server {
	if p.servers == nil {
		p.servers = map[string]*Server{}
	}
	v, ok := p.servers[server.Key]
	if ok {
		return v
	}
	p.servers[server.Key] = server
	return server
}

func (p *ServerPool) Close() {
	for _, server := range p.servers {
		server.Close()
	}
}
